### ISSUE TITLE
MapStore.storeAll() never called when objects updated with delay

### ISSUE THREAD

**Issue by dmitrymz** [hazelcast/hazelcast#7464]
I am implementing MapStore.storeAll() for our cluster to periodically backup the maps.
I configured the cluster nodes for write-behind feature, ie put 60sec for testing.
Ran the test, where updated one, two, three map values within a minute.
However, only store() method got called (3 times)
Expected behavior: storeAll() called once per minute with the full map contents.
please, advise.

```
   Hazelcast version: 3.5.5
   Cluster size: 2 nodes
   Number of the clients: 1
   Version of Java: 1.8.0_20
   Operating system: 3.13.0-77-generic #121-Ubuntu SMP Wed Jan 20 10:50:42 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux


```

**Comment by dmitrymz** [comment]
hazelcast.xml (partial)

```
   <properties>
        <property name="hazelcast.logging.type">slf4j</property>
        <property name="hazelcast.icmp.enabled">true</property>
    </properties>

    <network>
        <port auto-increment="false">5701</port>
        <join>
            <multicast enabled="false"/>
            <tcp-ip enabled="false"/>
            <aws enabled="true">
                <region>us-west-2</region>
                ...
            </aws>
        </join>
    </network>

    <serialization>
        <portable-version>0</portable-version>
        <portable-factories>
            <portable-factory factory-id="1">
                com.mz.retail.view.datagrid.model.MyPortableFactory
            </portable-factory>
        </portable-factories>
    </serialization>

    <map name="cont_*">
        <map-store enabled="true">
            <factory-class-name>com.mz.retail.view.datagrid.backup.MyMapStoreFactory</factory-class-name>
            <write-delay-seconds>60</write-delay-seconds>
        </map-store>
    </map>
</hazelcast>

```

**Comment by dmitrymz** [comment]
called to update value, here are the logs:
```
2016-02-04 18:55:35,629 INFO  [com.mz.retail.view.datagrid.backup.ContainerMapStore] ======== store
2016-02-04 18:55:39,629 INFO  [com.mz.retail.view.datagrid.backup.ContainerMapStore] ======== store
2016-02-04 18:55:42,629 INFO  [com.mz.retail.view.datagrid.backup.ContainerMapStore] ======== store
```

**Comment by dmitrymz** [comment]
this is a part of MapStore implementation:

public class MyMapStore implements MapStore<String, MZTypeDo> {

    private String mapName;
    private MapStoreDao mapStoreDao;

    public MyMapStore(String mapName, MapStoreDao mapStoreDao) {
        this.mapName = mapName;
        this.mapStoreDao = mapStoreDao;
        log.info("======== initialized");
    }

    @Override
    public void store(String key, MZTypeDo c) {
        log.info("======== store");
    }

    @Override
    public void storeAll(Map<String, MZTypeDo> org) {
        log.info("======== storeAll");

        mapStoreDao.store(mapName, org);
    }
...
}

**Comment by noctarius** [comment]
As far as I'm aware of MapStore::storeAll is only used whenever you call IMap::putAll.

**Comment by dmitrymz** [comment]
from the hazelcast.xml doc:

**Comment by noctarius** [comment]
Hm good catch, @ahmetmircik any idea about that?

**Comment by dmitrymz** [comment]
is it possible that       <write-batch-size> parameter is affecting this?
"The number of operations to be included in each batch processing round. Default value is 1."
if yes, is it valid to put some really large number to make sure all the records got persisted?

**Comment by dmitrymz** [comment]
answering my question above: no  write-batch-size has no effect.
tried the same test with write-batch-size=100000 - still singe store() method was called 3 times

**Comment by ahmetmircik** [comment]
Hi @dmitrymz,

Could you please share a simple reproducer test? It will be easier to talk on that.

**Comment by dmitrymz** [comment]
I don't have much of a test.
Our setup is a simple webservice storing values in Hazelcast.
On the cluster side we attached the MapStore with configuration mentioned above.

You can simply reproduce the issue just by having these 2 components.

**Comment by dmitrymz** [comment]
I did some basic debugging using setup above and hazelcast source code, and found that DefaultWriteBehindProcessor.callHandler() is always called with just one delayedEntry, therefore this check 
        if (size == 1 || !writeCoalescing) {
            return processEntriesOneByOne(delayedEntries, operationType);
        }
is always true - here is one by one call to store().

Expected: WriteBehindProcessor should accumulate updated records over period of time specified in 
write-delay-seconds, then flush them with storeAll() method or deleteAll() method. 
basically, we should never expect single store(), delete() methods to be invoked in write-behind mode in my understanding

**Comment by gurbuzali** [comment]
I've tried to reproduce your case here (with the latest master) but couldn't
https://gist.github.com/gurbuzali/34589488f6b5352d92be

**Comment by ahmetmircik** [comment]
@dmitrymz I see your point this is a bug. Batching should take place where it is possible. Thanks for reporting this.

**Comment by dmitrymz** [comment]
@gurbuzali I used 3.5.5 release from here https://github.com/hazelcast/hazelcast/commit/6805b8e41faa52f5c74d0edef6ec255b0f0b0d51

**Comment by dmitrymz** [comment]
seems like I found an issue, and a way to reproduce it:
add a simple delay of 1sec between updating the the objects, ie

```
while (true) {
 put(key, value);
 sleep(1000);
}

```
the problem is the way hazelcast server does the aggregates for puts. if they all come instantly in bulk then storeAll() is called. this explains why putAll() is working and put() is not

when puts are called with delays then, store() is called. the longer the delay - the more chance single store() to be called.

try experiment with delays 50ms, 100ms, 300ms, 500ms

**Comment by dmitrymz** [comment]
there is another subproblem that I found experimenting a loop above with no delays - storeAll() is split in multiple calls.

try this scenarion: update 100 elements with no delay.
you will see most likely two storeAll() calls within write-delay-seconds periods, ie

storeAll(29 elements)
storeAll(71 elements)

next time

storeAll(23 elements)
storeAll(77 elements)

looks like aggregation logic should be fixed:
1. to consolidate all single updates within write-delay-seconds
2. make just one call to storeAll() with consolidated values
3. never call single store(), and single delete() methods as per contract

**Comment by ahmetmircik** [comment]
seems the reason behind storeAll split is member-count and it is expected in this situation, since every member handles its own data.

**Comment by dmitrymz** [comment]
hmm. however, only one cluster node actually calls MapStore.storeAll(), isn't it?
if that's true, there is a consolidation of all updates within the cluster at this node - which make sense. 

Because updates have to be done in intervals of write-delay-seconds. Otherwise, write-delay-seconds don't make sense if all the nodes start storing updates in their different intervals, invoking local MapStore.storeAll() at different times...
